src/System.Collections/src/System/Collections/Generic/HashSet.cs

```diff
@@ -1711,6 +1716,95 @@ internal static bool HashSetEquals(HashSet<T> set1, HashSet<T> set2, IEqualityCo
+        /// to specified comparer.
+        /// Because items are hashed according to a specific equality comparer, we have to resort
+        /// to building a new set if neither use that equality comparer.
+        /// </summary>
+        /// <param name="set1">The first set to compare.</param>
+        /// <param name="set2">The second set to compare.</param>
+        /// <param name="comparer">The comparer for comparing elements.</param>
+        /// <returns><c>true</c> if the two sets are equivalent, <c>false</c> otherwise.</returns>
+        internal static bool HashSetEquals(HashSet<T> set1, HashSet<T> set2, IEqualityComparer<T> comparer)
+        {
+            if (set1 == set2)
+            {
+                return true;
+            }
+
+            if (set1 == null | set2 == null)
+            {
+                return false;
+            }
+
+            int count1 = set1.Count;
+            int count2 = set2.Count;
+            if (count1 == 0)
+            {
+                return count2 == 0;
+            }
+
+            if (count2 == 0)
+            {
+                return false;
+            }
+
+            if (comparer.Equals(set1._comparer))
+            {
+                if (comparer.Equals(set2._comparer))
+                {
+                    // Since both sets are using our comparer as their comparer, we can test for set-equality.
+                    return set1.Count == set2.Count && set1.ContainsAllElements(set2);
+                }
+
+                if (count1 > count2)
+                {
+                    // Filtering set2 according to comparer can result in fewer than set2.Count elements,
+                    // but never more, so short-circuit if set2.Count is less than set1.Count.
+                    return false;
+                }
+            }
+            else
+            {
+                // We need a starting point of a set according to the comparer, and set1 doesn't meet
+                // that criteria. If set2 does, swap to use it.
+                if (comparer.Equals(set2._comparer))
+                {
+                    if (count1 < count2)
+                    {
+                        return false;
+                    }
+
+                    HashSet<T> swapTemp = set1;
+                    set1 = set2;
+                    set2 = swapTemp;
+                    count1 = count2;
+                }
+                else
+                {
+                    // Niether set use our comparer, so create a new one with the contents of set1.
+                    set1 = new HashSet<T>(set1, comparer);
+
+                    if ((count1 = set1.Count) > count2)
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            HashSet<T> seen = new HashSet<T>(count1 + 1, comparer);
+            foreach (T item in set2)
+            {
+                if (seen.Add(item) && !set1.Contains(item))
```

Comment:
Everywhere else I'm aware of in the framework, when there are overloads that differ only by whether an EqualityComparer is taken, the behavior without the comparer is the same as with a comparer and EqualityComparer.Default.  That's not happening here.  I understand the arguments you're making, but I think this aspect is confusing.  Maybe I'm alone in that.
cc: @terrajobst, @bartonjs